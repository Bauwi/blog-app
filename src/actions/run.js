import database from '../firebase/firebase';
import { db } from '../firebase/firebase';
import * as firebase from 'firebase';

export const addPostToRun = post => ({
  type: 'ADD_POST_TO_RUN',
  post
});

export const startAddPostToRun = post => (dispatch, getState) => {
  const { uid } = getState().auth;
  console.log(post);
  db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .add(post)
    .then((doc) => {
      dispatch(addPostToRun({
        DBid: doc.id,
        ...post
      }));
    });
};

export const removePostToRun = id => ({
  type: 'REMOVE_POST_TO_RUN',
  id
});

export const startRemovePostToRun = (id, DBid) => (dispatch, getState) => {
  const { uid } = getState().auth;
  console.log(DBid);
  return db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .doc(`${DBid}`)
    .delete()
    .then(dispatch(removePostToRun(id)));
};

export const setRunPosts = posts => ({
  type: 'SET_RUN_POSTS',
  posts
});

export const startSetRunPosts = () => (dispatch, getState) => {
  const { uid } = getState().auth;
  return db
    .collection('runs')
    .doc(uid)
    .collection('UserRuns')
    .get()
    .then((querySnapshot) => {
      const run = [];
      querySnapshot.forEach((doc) => {
        run.push({
          DBid: doc.id,
          ...doc.data()
        });
      });
      dispatch(setRunPosts(run));
    });
};
