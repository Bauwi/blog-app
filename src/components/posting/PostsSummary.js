import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';

import selectPosts from '../../selectors/posts';

export const PostsSummary = ({ selectedPosts, totalUserPosts }) => {
  const totalInMinutes = selectedPosts.reduce((total, post) => (total += post.readingTime), 0);
  const hiddenPostsNum = totalUserPosts - selectedPosts.length;
  return (
    <div className="page-header__container">
      <div className="content-container">
        <div className="page-header__actions">
          <Link className="button" to="/create">
            <i className="fa fa-pencil" /> New Post
          </Link>
        </div>
        <h1 className="page-header__title">
          {selectedPosts.length} post{selectedPosts.length > 1 ? 's' : ''}. {totalInMinutes} minute{totalInMinutes > 1 ? 's' : ''}{' '}
          read.
        </h1>
        <h3 className="page-header__subtitle">{hiddenPostsNum} hidden posts</h3>
      </div>
    </div>
  );
};

PostsSummary.propTypes = {
  selectedPosts: PropTypes.arrayOf(PropTypes.object).isRequired,
  totalUserPosts: PropTypes.number.isRequired
};

const mapStateToProps = state => ({
  selectedPosts: selectPosts(state.posts.posts, state.filters),
  totalUserPosts: state.users.preferences.numberOfPosts
});

export default connect(mapStateToProps)(PostsSummary);
