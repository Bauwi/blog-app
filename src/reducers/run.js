const runReducerDefaultState = {
  isLoading: true, posts: [], hasErrored: false, current: {}
};

const runReducer = (state = runReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_POST_TO_RUN':
      return {
        ...state,
        posts: [...state.posts, action.post]
      };
    case 'REMOVE_POST_TO_RUN':
      return {
        ...state,
        posts: state.posts.filter(post => action.id !== post.content.id)
      };
    case 'SET_RUN_POSTS':
      return {
        ...state,
        posts: action.posts
      };
    case 'RUN_IS_LOADING':
      return {
        ...state,
        isLoading: action.isLoading
      };
    case 'RUN_HAS_ERRORED':
      return {
        ...state,
        hasErrored: action.hasErrored
      };
    case 'SET_CURRENT_POST_RUN':
      return {
        ...state,
        current: state.posts.find(post => post.content.id === action.id)
      };
    case 'UPDATE_RUN_POST_TO_ALREADY_READ':
      return {
        ...state,
        current: { ...state.current, state: 'read' },
        posts: state.posts.map((post) => {
          if (post.content.id === action.id) {
            return {
              ...post,
              state: 'read'
            };
          }
          return post;
        })
      };
    case 'UPDATE_RUN_POST_TO_NOT_ALREADY_READ':
      return {
        ...state,
        posts: state.posts.map((post) => {
          if (post.content.id === action.id) {
            return {
              ...post,
              state: 'unread'
            };
          }
          return post;
        })
      };
    case 'RESET_RUN':
      return {
        posts: []
      };
    case 'CLEAN_RUN':
      return {
        ...state,
        posts: state.posts.filter(post => post.state === 'unread')
      };
    default:
      return state;
  }
};

export default runReducer;
