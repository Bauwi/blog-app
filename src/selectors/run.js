const getRunPosts = (posts = [], run) => run.map(read => posts.find(post => post.id === read.id));
export default getRunPosts;
