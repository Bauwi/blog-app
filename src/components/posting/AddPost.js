import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import PostForm from './PostForm';
import { startAddPost } from '../../actions/posts';
import { startUpdateUser } from '../../actions/users';

export class AddPost extends Component {
  onSubmit = post => {
    this.props.startAddPost(post);
    this.props.startUpdateUser(undefined, { username: post.author });
    this.props.history.push('/');
  };

  render() {
    return (
      <div>
        <div className="page-header">
          <header className="content-container">
            <Link to="/dashboard">Back to dashboard</Link>
          </header>
        </div>
        <PostForm onSubmit={this.onSubmit} context="add" author={this.props.user} />
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  startAddPost: post => dispatch(startAddPost(post)),
  startUpdateUser: (id, updates) => dispatch(startUpdateUser(id, updates))
});

const mapStateToProps = state => ({
  user: state.users.preferences
});
export default connect(mapStateToProps, mapDispatchToProps)(AddPost);
