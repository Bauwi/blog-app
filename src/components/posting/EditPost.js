import React, { Component } from 'react';
import { connect } from 'react-redux';
import { startEditPost, startRemovePost } from '../../actions/posts';

import PostForm from './PostForm';

export class EditPost extends Component {
  onSubmit = post => {
    this.props.startEditPost(this.props.post.id, post);
    this.props.history.push('/dashboard');
  };
  onRemovePost = () => {
    this.props.startRemovePost(this.props.post.id);
    this.props.history.push('/dashboard');
  };

  render() {
    return (
      <div>
        <PostForm onSubmit={this.onSubmit} post={this.props.post} />
        <button onClick={this.onRemovePost}>Remove Post</button>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => {
  return {
    post: state.posts.find(post => {
      return post.id === props.match.params.id;
    })
  };
};

const mapDispatchToProps = (dispatch, props) => ({
  startEditPost: (id, post) => dispatch(startEditPost(id, post)),
  startRemovePost: id => dispatch(startRemovePost(id))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditPost);
