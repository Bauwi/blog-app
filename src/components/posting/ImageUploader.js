import React, { Component } from 'react';
import { connect } from 'react-redux';
import { uploadImage } from '../../actions/posts';

export class ImageUploader extends Component {
  handleChange = e => {
    const image = e.target.files[0];
    this.props.uploadImage(image);
  };

  render() {
    return (
      <div>
        <progress value="0" max="100">
          0%
        </progress>
        <input type="file" onChange={this.handleChange} />
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  uploadImage: image => dispatch(uploadImage(image))
});

export default connect(undefined, mapDispatchToProps)(ImageUploader);
