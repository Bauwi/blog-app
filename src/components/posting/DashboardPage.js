import React, { Component } from 'react';
import { connect } from 'react-redux';

import Header from '../header/Header';
import PostsList from './PostsList';
import PostsSummary from './PostsSummary';
import UserCard from '../UserCard';

export class DashboardPage extends Component {
  state = { goTopIcon: false };
  render() {
    return (
      <div className="content-container">
        <UserCard author={this.props.user} />
        <PostsSummary />
        <PostsList />
      </div>
    );
  }
}

const mapStateToProps = state => ({
  user: state.users.preferences
});

export default connect(mapStateToProps)(DashboardPage);
