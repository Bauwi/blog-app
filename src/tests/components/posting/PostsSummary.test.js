import React from 'react';
import { shallow } from 'enzyme';
import { PostsSummary } from '../../../components/posting/PostsSummary';
import run from '../../fixtures/run';

const posts = run.posts.map(post => post.content);

test('should render PostsSummary component properly with no post provided', () => {
  const wrapper = shallow(<PostsSummary totalUserPosts={0} selectedPosts={[]} />);
  expect(wrapper).toMatchSnapshot();
});

test('should render PostsSummary component properly with multiple posts provided', () => {
  const wrapper = shallow(<PostsSummary totalUserPosts={3} selectedPosts={posts} />);
  expect(wrapper).toMatchSnapshot();
});
