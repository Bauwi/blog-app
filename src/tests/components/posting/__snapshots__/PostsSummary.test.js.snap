// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render PostsSummary component properly with multiple posts provided 1`] = `
<div
  className="page-header__container"
>
  <div
    className="content-container"
  >
    <div
      className="page-header__content"
    >
      <div>
        <h1
          className="page-header__title"
        >
          Dashboard
        </h1>
        <p
          className="page-header__title"
        >
          3
           post
          s
           totalling
           
          21
           minute
          s
           read.
        </p>
        <p
          className="page-header__subtitle"
        >
          0
           hidden posts
        </p>
      </div>
      <div
        className="page-header__actions"
      >
        <Link
          className="button button--icon"
          replace={false}
          to="/create"
        >
          <i
            className="fa fa-pencil-square-o"
          />
        </Link>
      </div>
    </div>
    <PostListFilters />
  </div>
</div>
`;

exports[`should render PostsSummary component properly with no post provided 1`] = `
<div
  className="page-header__container"
>
  <div
    className="content-container"
  >
    <div
      className="page-header__content"
    >
      <div>
        <h1
          className="page-header__title"
        >
          Dashboard
        </h1>
        <p
          className="page-header__title"
        >
          0
           post
           totalling
           
          0
           minute
           read.
        </p>
        <p
          className="page-header__subtitle"
        >
          0
           hidden posts
        </p>
      </div>
      <div
        className="page-header__actions"
      >
        <Link
          className="button button--icon"
          replace={false}
          to="/create"
        >
          <i
            className="fa fa-pencil-square-o"
          />
        </Link>
      </div>
    </div>
    <PostListFilters />
  </div>
</div>
`;
